client/cli/src/lib.rs

```diff
@@ -161,637 +144,227 @@ fn base_path(
-				eprintln!("{:?} did not exist.", &db_path);
-				Ok(())
-			},
-			Result::Err(err) => Result::Err(err.into())
-		}
-	}
-}
+	config.chain_spec = Some(load_spec(shared_params, spec_factory)?);
+	config.config_dir = Some(base_path(shared_params, version));
+	config.impl_commit = version.commit;
+	config.impl_version = version.version;
 
-/// Command ready to revert the chain.
-pub struct ParseAndPrepareRevert<'a> {
-	params: RevertCmd,
-	version: &'a VersionInfo,
-}
+	config.database = DatabaseConfig::Path {
+		path: config
+			.in_chain_config_dir(DEFAULT_DB_CONFIG_PATH)
+			.expect("We provided a base_path/config_dir."),
+		cache_size: None,
+	};
 
-impl<'a> ParseAndPrepareRevert<'a> {
-	/// Runs the command and reverts the chain.
-	pub fn run_with_builder<C, G, E, F, B, S>(
-		self,
-		builder: F,
-		spec_factory: S
-	) -> error::Result<()> where
-		S: FnOnce(&str) -> Result<Option<ChainSpec<G, E>>, String>,
-		F: FnOnce(Configuration<C, G, E>) -> Result<B, error::Error>,
-		B: ServiceBuilderCommand,
-		<<<<B as ServiceBuilderCommand>::Block as BlockT>::Header as HeaderT>
-			::Number as FromStr>::Err: Debug,
-		C: Default,
-		G: RuntimeGenesis,
-		E: ChainSpecExtension,
-	{
-		let mut config = create_config_with_db_path(
-			spec_factory,
-			&self.params.shared_params,
-			self.version,
-			None,
-		)?;
-		fill_config_keystore_in_memory(&mut config)?;
+	config.network.boot_nodes = config.expect_chain_spec().boot_nodes().to_vec();
+	config.telemetry_endpoints = config.expect_chain_spec().telemetry_endpoints().clone();
 
-		let blocks = self.params.num.parse()?;
-		builder(config)?.revert_chain(blocks)?;
-		Ok(())
-	}
+	Ok(())
 }
 
-/// Create a `NodeKeyConfig` from the given `NodeKeyParams` in the context
-/// of an optional network config storage directory.
-fn node_key_config<P>(params: NodeKeyParams, net_config_dir: &Option<P>)
-	-> error::Result<NodeKeyConfig>
+/// Run the node
+///
+/// Run the light node if the role is "light", otherwise run the full node.
+pub fn run_node<G, E, FNL, FNF, SL, SF>(
+	config: Configuration<G, E>,
+	new_light: FNL,
+	new_full: FNF,
+	version: &VersionInfo,
+) -> error::Result<()>
 where
-	P: AsRef<Path>
+	FNL: FnOnce(Configuration<G, E>) -> Result<SL, sc_service::error::Error>,
+	FNF: FnOnce(Configuration<G, E>) -> Result<SF, sc_service::error::Error>,
+	G: RuntimeGenesis,
+	E: ChainSpecExtension,
+	SL: AbstractService + Unpin,
+	SF: AbstractService + Unpin,
 {
-	match params.node_key_type {
-		NodeKeyType::Ed25519 =>
-			params.node_key.as_ref().map(parse_ed25519_secret).unwrap_or_else(||
-				Ok(params.node_key_file
-					.or_else(|| net_config_file(net_config_dir, NODE_KEY_ED25519_FILE))
-					.map(sc_network::config::Secret::File)
-					.unwrap_or(sc_network::config::Secret::New)))
-				.map(NodeKeyConfig::Ed25519)
+	info!("{}", version.name);
+	info!("  version {}", config.full_version());
+	info!("  by {}, 2017-2020", version.author);
```

Comment:
When a project started in 2020, this wouldn't be correct :P 

And what happens in 2021?

Reply:
The code moved. There is no support for the date in `VersionInfo`. I see only 2 possibilities to fix this:

1. delete `run_node` and let the caller do this manually: this is okay since `run_node` doesn't do  much anymore, only starts the service if light or full depending on the role
2. delete the date from the info

What do you want? (or you have another idea?)

Reply:
option 2 applied

Reply:
3. Add `copyright_start_year` or similar to `VersionInfo`. The second year can just be `current_year`.

Reply:
OK. but to get the current year I will have to introduce chrono

Reply:
done